File: qlang-interpreter/src/qlang/runtime/interpreter.ts
import type {
    ArrayExpression,
    AssignmentExpression,
    BinaryExpression,
    BlockStatement,
    BooleanLiteral,
    CallExpression,
    ForStatement,
    FunctionStatement,
    Identifier,
    IfStatement,
    MemberExpression,
    NumericLiteral,
    PrintStatement,
    Program,
    ReadExpression,
    ReturnStatement,
    Statement,
    StringLiteral,
    UnaryExpression,
    VariableDeclarationStatement,
    WhileStatement,
} from '../ast'
import {
    type AlgebraicValue,
    ArrayValue,
    FunctionValue,
    MK_ALGEBRAIC,
    MK_BOOLEAN,
    MK_BREAK,
    MK_CONTINUE,
    MK_FUNCTION,
    MK_NULL,
    MK_NUMBER,
    MK_RETURN,
    MK_STRING,
    NumberValue,
    type RuntimeValue,
} from './values'
import Environment from './environment'
import { Std } from './std'
import { QFunction } from './callable'

export default class Interpreter {

    private env: Environment
    private stdOut: Std
    private stdErr: Std

    public get StdOut(): Std { return this.stdOut }
    public get StdErr(): Std { return this.stdErr }

    constructor(env: Environment, stdOut: Std, stdErr: Std) {
        this.env = env
        this.stdOut = stdOut
        this.stdErr = stdErr
    }

    public evaluate(astNode: Statement): RuntimeValue {
        switch (astNode.kind) {
            case 'Program':
                return this.evaluateProgram(astNode as Program)
            case 'VariableDeclarationStatement':
                return this.evaluateVariableDeclaration(astNode as VariableDeclarationStatement)
            case 'PrintStatement':
                return this.evaluatePrintStatement(astNode as PrintStatement)
            case 'BlockStatement':
                return this.evaluateBlockStatement(astNode as BlockStatement, this.env)
            case 'IfStatement':
                return this.evaluateIfStatement(astNode as IfStatement)
            case 'WhileStatement':
                return this.evaluateWhileStatement(astNode as WhileStatement)
            case 'ForStatement':
                return this.evaluateForStatement(astNode as ForStatement)
            case 'FunctionStatement':
                return this.evaluateFunctionDeclaration(astNode as FunctionStatement)
            default:
                return this.evaluateExpression(astNode)
        }
    }

    private evaluateProgram(program: Program): RuntimeValue {
        let lastEvaluated: RuntimeValue = MK_NULL()

        for (const statement of program.body) {
            lastEvaluated = this.evaluate(statement)
        }
        return lastEvaluated
    }

    private evaluateVariableDeclaration(variableDeclaration: VariableDeclarationStatement): RuntimeValue {
        if (variableDeclaration.value) {
            const value = this.evaluate(variableDeclaration.value)
            this.env.declareVariable(variableDeclaration.identifier, value)
            return value
        }
        this.env.declareVariable(variableDeclaration.identifier, MK_NULL())
        return MK_NULL()
    }

    private evaluatePrintStatement(printStatement: PrintStatement): RuntimeValue {
        const statement = this.evaluate(printStatement.value)
        this.stdOut.print(this.toString(statement))

        return MK_NULL()
    }

    private toString(statement: RuntimeValue): string {
        if (!('value' in statement)) {
            return statement.type
        }

        if (!Array.isArray(statement.value)) {
            return statement.value?.toString() ?? 'rien'
        }

        return this.toStringArray(statement as ArrayValue)
    }

    private toStringArray(array: ArrayValue): string {
        const elements = array.value.map((element) => this.toString(element))
        return `[${elements.join(', ')}]`
    }

    public evaluateBlockStatement(blockStatement: BlockStatement, environment: Environment): RuntimeValue {
        const previousEnv = this.env
        this.env = environment

        let lastEvaluated: RuntimeValue = MK_NULL()

        for (const statement of blockStatement.body) {
            if (statement.kind === 'BreakStatement') {
                this.env = previousEnv
                return MK_BREAK()
            }
            if (statement.kind === 'ContinueStatement') {
                this.env = previousEnv
                return MK_CONTINUE()
            }
            if (statement.kind === 'ReturnStatement') {
                const result = this.evaluate((statement as ReturnStatement).value) as AlgebraicValue
                this.env = previousEnv
                return MK_RETURN(result)
            }
            lastEvaluated = this.evaluate(statement)
            if (lastEvaluated.type === 'break' || lastEvaluated.type === 'continue' || lastEvaluated.type === 'return') {
                this.env = previousEnv
                return lastEvaluated
            }
        }
        return lastEvaluated
    }

    private evaluateIfStatement(ifStatement: IfStatement): RuntimeValue {
        const condition = this.evaluate(ifStatement.condition) as AlgebraicValue

        if (condition.value) {
            this.env = new Environment(this.env)
            const result = this.evaluate(ifStatement.thenBranch)

            this.env = this.env.Parent!
            return result
        }

        if (ifStatement.elseBranch) {
            this.env = new Environment(this.env)
            const result = this.evaluate(ifStatement.elseBranch)

            this.env = this.env.Parent!
            return result
        }

        return MK_NULL()
    }

    private evaluateWhileStatement(whileStatement: WhileStatement): RuntimeValue {
        this.env = new Environment(this.env)
        while ((this.evaluate(whileStatement.condition) as AlgebraicValue).value) {
            const result = this.evaluate(whileStatement.body)
            if (result.type === 'break') {
                break
            }
            if (result.type === 'return') {
                this.env = this.env.Parent!
                return result
            }
        }

        this.env = this.env.Parent!
        return MK_NULL()
    }

    private evaluateForStatement(forStatement: ForStatement): RuntimeValue {
        this.env = new Environment(this.env)
        if (this.env.resolve(forStatement.identifier, false) === null) {
            this.env.declareVariable(forStatement.identifier, MK_NULL())
        }
        this.env.assignVariable(forStatement.identifier, this.evaluate(forStatement.from))

        while ((this.evaluate(forStatement.until) as AlgebraicValue).value) {
            const result = this.evaluate(forStatement.body)
            if (result.type === 'break') {
                break
            }
            if (result.type === 'return') {
                this.env = this.env.Parent!
                return result
            }
            this.evaluate(forStatement.step)
        }
        this.env = this.env.Parent!

        return MK_NULL()
    }

    private evaluateFunctionDeclaration(functionStatement: FunctionStatement): RuntimeValue {
        const qfunction = new QFunction(functionStatement, this.env)
        const name = qfunction.Name

        const envFound = this.env.resolve(name, false)
        if (envFound === null) {
            this.env.declareVariable(name, MK_NULL())
        }
        return this.env.assignVariable(name, MK_FUNCTION(qfunction))
    }

    private evaluateExpression(expression: Statement): RuntimeValue {
        switch (expression.kind) {
            case 'AssignmentExpression':
                return this.evaluateAssignmentExpression(expression as AssignmentExpression)
            case 'UnaryExpression':
                return this.evaluateUnaryExpression(expression as UnaryExpression)
            case 'BinaryExpression':
                return this.evaluateBinaryExpression(expression as BinaryExpression)
            case 'ArrayExpression':
                return this.evaluateArrayExpression(expression as ArrayExpression)
            case 'MemberExpression':
                return this.evaluateMemberExpression(expression as MemberExpression)
            case 'CallExpression':
                return this.evaluateCallExpression(expression as CallExpression)
            case 'ReadExpression':
                return this.evaluateReadExpression(expression as ReadExpression)
            case 'Identifier':
                return this.evaluateIdentifier(expression as Identifier)
            case 'NumericLiteral':
                return MK_NUMBER((expression as NumericLiteral).value)
            case 'StringLiteral':
                return MK_STRING((expression as StringLiteral).value)
            case 'NullLiteral':
                return MK_NULL()
            case 'BooleanLiteral':
                return MK_BOOLEAN((expression as BooleanLiteral).value)
            default:
                // TODO: throw error
                throw new Error()
        }
    }

    private evaluateAssignmentExpression(node: AssignmentExpression): RuntimeValue {
        if (node.assignment.kind === 'Identifier') {
            const identifier = (node.assignment as Identifier).name
            return this.env.assignVariable(identifier, this.evaluate(node.value))
        }

        if (node.assignment.kind === 'MemberExpression') {
            const memberExpression = node.assignment as MemberExpression
            const object = this.evaluateExpression(memberExpression.object) as AlgebraicValue
            if (object.type !== 'array') {
                // TODO: throw error
                throw new Error()
            }

            const array = object as ArrayValue
            if (memberExpression.property === null) {
                const value = this.attemptAlgebraicValue(this.evaluateExpression(node.value))
                array.value.push(value)

                return value
            }

            const property = this.evaluateExpression(memberExpression.property)
            if (property.type !== 'number') {
                // TODO: throw error
                throw new Error()
            }

            const index = (property as NumberValue).value
            if (index < 0 || index >= array.value.length) {
                // TODO: throw error
                throw new Error()
            }

            const value = this.attemptAlgebraicValue(this.evaluateExpression(node.value))
            return array.value[index] = value
        }

        // TODO: throw error
        throw new Error()
    }

    private evaluateUnaryExpression(unaryExpr: UnaryExpression): RuntimeValue {
        const argument = this.evaluate(unaryExpr.value)

        if ('value' in argument) {
            if (unaryExpr.operator === '-') {
                return MK_NUMBER(-(argument.value as number))
            } else if (unaryExpr.operator === 'non') {
                return MK_BOOLEAN(!(argument.value as boolean))
            }
        }

        return MK_NULL()
    }

    private evaluateBinaryExpression(binaryExpr: BinaryExpression): RuntimeValue {
        if (this.isLogicalOperator(binaryExpr.operator)) {
            return this.evaluateLogicalBinaryExpression(binaryExpr)
        }

        const leftHandSide = this.evaluate(binaryExpr.left)
        const rightHandSide = this.evaluate(binaryExpr.right)

        return this.evaluateAlgebraicBinaryExpression(
            binaryExpr.operator,
            leftHandSide as AlgebraicValue,
            rightHandSide as AlgebraicValue,
        )
    }

    private isLogicalOperator(operator: string): boolean {
        return ['et', 'ou', '==', '!=', '<', '<=', '>', '>='].includes(operator)
    }

    private evaluateArrayExpression(arrayExpression: ArrayExpression): RuntimeValue {
        return {
            type: 'array',
            value: arrayExpression.elements.map((element) => this.evaluate(element))
        } as ArrayValue
    }

    private evaluateMemberExpression(memberExpression: MemberExpression): RuntimeValue {
        const object = this.evaluateExpression(memberExpression.object)
        if (object.type !== 'array') {
            // TODO: throw error
            throw new Error()
        }

        const array = object as ArrayValue
        if (memberExpression.property === null) {
            // TODO: throw error
            throw new Error()
        }

        const property = this.evaluateExpression(memberExpression.property)
        if (property.type !== 'number') {
            // TODO: throw error
            throw new Error()
        }

        const index = (property as NumberValue).value
        if (index < 0 || index >= array.value.length) {
            // TODO: throw error
            throw new Error()
        }

        return array.value[index]
    }

    private evaluateCallExpression(callExpression: CallExpression): RuntimeValue {
        const expression = this.evaluateExpression(callExpression.callee)
        if (expression.type !== 'function') {
            // TODO: throw error
            throw new Error()
        }

        const callee = expression as FunctionValue
        if (callExpression.arguments.length !== callee.value.Arity) {
            // TODO: throw error
            throw new Error()
        }

        const args = callExpression.arguments.map(
            (argument) => this.attemptAlgebraicValue(this.evaluate(argument)),
        )

        const currentValue = callee.value.call(this, args)
        return currentValue.type === 'return'
            ? MK_ALGEBRAIC(currentValue.value)
            : currentValue
    }

    private evaluateReadExpression(readExpr: ReadExpression): RuntimeValue {
        const algebraicValue = this.evaluate(readExpr.value) as AlgebraicValue
        const message = algebraicValue?.value
        const result = prompt(message?.toString() ?? 'Input')

        return result === null ? MK_NULL() : MK_STRING(result)
    }

    private evaluateIdentifier(identifier: Identifier): RuntimeValue {
        return this.env.lookupVariable(identifier.name)
    }

    private evaluateLogicalBinaryExpression(binaryExpr: BinaryExpression): AlgebraicValue {
        const leftHandSide = this.evaluate(binaryExpr.left) as AlgebraicValue
        const operator = binaryExpr.operator

        if (operator === 'et') {
            if (leftHandSide.type === 'boolean' && leftHandSide.value === false) {
                return MK_BOOLEAN(false)
            }
            return MK_BOOLEAN((this.evaluate(binaryExpr.right) as AlgebraicValue).value as boolean)
        }
        if (operator === 'ou') {
            if (leftHandSide.type === 'boolean' && leftHandSide.value === true) {
                return MK_BOOLEAN(true)
            }
            return MK_BOOLEAN((this.evaluate(binaryExpr.right) as AlgebraicValue).value as boolean)
        }

        const right = this.evaluate(binaryExpr.right) as AlgebraicValue
        if (operator === '==') {
            return MK_BOOLEAN(leftHandSide.value === right.value)
        }
        if (operator === '!=') {
            return MK_BOOLEAN(leftHandSide.value !== right.value)
        }
        if (operator === '<') {
            return MK_BOOLEAN(leftHandSide.value! < right.value!)
        }
        if (operator === '<=') {
            return MK_BOOLEAN(leftHandSide.value! <= right.value!)
        }
        if (operator === '>') {
            return MK_BOOLEAN(leftHandSide.value! > right.value!)
        }
        if (operator === '>=') {
            return MK_BOOLEAN(leftHandSide.value! >= right.value!)
        }

        // TODO: throw error
        throw new Error()
    }

    private evaluateAlgebraicBinaryExpression(
        operator: string,
        left: AlgebraicValue,
        right: AlgebraicValue,
    ): AlgebraicValue {
        if ((left.type === 'string' || right.type === 'string') && operator !== '+') {
            // TODO: throw error
            throw new Error()
        }
        if ((left.type === 'string' || right.type === 'string')) {
            return MK_STRING((left.value as string) + (right.value as string))
        }
        if (operator === '+') {
            return MK_NUMBER((left.value as number) + (right.value as number))
        } else if (operator === '-') {
            return MK_NUMBER((left.value as number) - (right.value as number))
        } else if (operator === '*') {
            return MK_NUMBER((left.value as number) * (right.value as number))
        } else if (operator === '/') {
            return MK_NUMBER((left.value as number) / (right.value as number))
        } else if (operator === '%') {
            return MK_NUMBER((left.value as number) % (right.value as number))
        }

        // TODO: throw error
        throw new Error()
    }

    private attemptAlgebraicValue(value: RuntimeValue): AlgebraicValue {
        if ('value' in value) {
            return value as AlgebraicValue
        }

        // TODO: throw error
        throw new Error()
    }
}

File: interpreter/src/interpreter.py
from typing import Self

from src.nodes.merged import BlockStatement
from src.nodes.merged import FunctionStatement
from src.nodes.merged import Statement
from src.nodes.node_type import NodeType
from src.nodes.statement import ForStatement
from src.nodes.statement import IfStatement
from src.nodes.statement import PrintStatement
from src.nodes.statement import Program
from src.nodes.statement import ReturnStatement
from src.nodes.statement import VariableDeclarationStatement
from src.nodes.statement import WhileStatement
from src.runtime.callable import QFunction
from src.runtime.environment import Environment
from src.runtime.std import Std
from src.runtime.values import MK_BREAK
from src.runtime.values import MK_CONTINUE
from src.runtime.values import MK_FUNCTION
from src.runtime.values import MK_NULL
from src.runtime.values import MK_RETURN
from src.runtime.values import AlgebraicType
from src.runtime.values import AlgebraicValue
from src.runtime.values import RuntimeValue


class Interpreter:

    env: Environment
    std_out: Std
    std_err: Std

    def __init__(self: Self, env: Environment, std_out: Std, std_err: Std) -> None:
        self.env = env
        self.std_out = std_out
        self.std_err = std_err

    def evaluate(self: Self, ast_node: Statement) -> RuntimeValue:
        if ast_node.kind == NodeType.PROGAM:
            return self.__evaluate_program_statement(ast_node, self.env)  # type: ignore
        if ast_node.kind == NodeType.BLOCK_STATEMENT:
            return self.evaluate_block_statement(ast_node, self.env)  # type: ignore
        if ast_node.kind == NodeType.VARIABLE_DECLARATION_STATEMENT:
            return self.__evaluate_variable_declaration_statement(ast_node)  # type: ignore
        if ast_node.kind == NodeType.PRINT_STATEMENT:
            return self.__evaluate_print_statement(ast_node)  # type: ignore

        if ast_node.kind == NodeType.IF_STATEMENT:
            return self.__evaluate_if_statement(ast_node)  # type: ignore
        if ast_node.kind == NodeType.WHILE_STATEMENT:
            return self.__evaluate_while_statement(ast_node)  # type: ignore
        if ast_node.kind == NodeType.FOR_STATEMENT:
            return self.__evaluate_for_statement(ast_node)  # type: ignore
        if ast_node.kind == NodeType.FUNCTION_STATEMENT:
            return self.__evaluate_function_statement(ast_node)  # type: ignore

        return self.__evaluate_expression(ast_node)

    def evaluate_block_statement(
        self: Self, block_statement: BlockStatement, environment: Environment
    ) -> RuntimeValue:
        previous_env = self.env
        self.env = environment

        last_evaluated: RuntimeValue = MK_NULL()
        for statement in block_statement.body:
            if statement.kind == NodeType.BREAK_STATEMENT:
                self.env = previous_env
                return MK_BREAK()
            if statement.kind == NodeType.CONTINUE_STATEMENT:
                self.env = previous_env
                return MK_CONTINUE()
            if statement.kind == NodeType.RETURN_STATEMENT:
                stmt: ReturnStatement = statement  # type: ignore
                result: AlgebraicValue = self.evaluate(stmt.value)  # type: ignore

                self.env = previous_env
                return MK_RETURN(result)

            last_evaluated = self.evaluate(statement)
            if (
                last_evaluated.type == AlgebraicType.BREAK
                or last_evaluated.type == AlgebraicType.CONTINUE
                or last_evaluated.type == AlgebraicType.RETURN
            ):
                self.env = previous_env
                return last_evaluated

        return last_evaluated

    def __evaluate_program_statement(
        self: Self, program_statement: Program
    ) -> RuntimeValue:
        last_evaluated: RuntimeValue = MK_NULL()

        for statement in program_statement.body:
            last_evaluated = self.evaluate(statement)
            if (
                last_evaluated.type == AlgebraicType.BREAK
                or last_evaluated.type == AlgebraicType.CONTINUE
                or last_evaluated.type == AlgebraicType.RETURN
            ):
                return last_evaluated

        return last_evaluated

    def __evaluate_variable_declaration_statement(
        self: Self, variable_declaration_statement: VariableDeclarationStatement
    ) -> RuntimeValue:
        if variable_declaration_statement.value is not None:
            value = self.evaluate(variable_declaration_statement.value)
            self.env.declareVariable(variable_declaration_statement.identifier, value)
            return value

        self.env.declareVariable(variable_declaration_statement.identifier, MK_NULL())
        return MK_NULL()

    def __evaluate_print_statement(
        self: Self, print_statement: PrintStatement
    ) -> RuntimeValue:
        statement = self.evaluate(print_statement.value)
        self.std_out.print(self.__runtime_string(statement))
        return MK_NULL()

    def __runtime_string(self: Self, statement: RuntimeValue) -> str:
        if not hasattr(statement, "value"):
            return str(statement.type)

        if statement.type != AlgebraicType.ARRAY:
            return str(statement.value)

        arr = statement.value
        return f"[{', '.join([self.__runtime_string(val) for val in arr])}]"

    def __evaluate_if_statement(self: Self, if_statement: IfStatement) -> RuntimeValue:
        condition = self.__to_algebraic_value(self.evaluate(if_statement.condition))

        if condition.value:
            self.env = Environment(self.env)
            result = self.evaluate(if_statement.then_branch)

            self.env = self.env.parent  # type: ignore
            return result

        if if_statement.else_branch is not None:
            self.env = Environment(self.env)
            result = self.evaluate(if_statement.else_branch)

            self.env = self.env.parent  # type: ignore
            return result

        return MK_NULL()

    def __evaluate_while_statement(
        self: Self, while_statement: WhileStatement
    ) -> RuntimeValue:
        self.env = Environment(self.env)

        while self.__to_algebraic_value(self.evaluate(while_statement.condition)).value:
            result = self.evaluate(while_statement.body)
            if result.type == AlgebraicType.BREAK:
                break
            if result.type == AlgebraicType.RETURN:
                self.env = self.env.parent  # type: ignore
                return result

        self.env = self.env.parent  # type: ignore
        return MK_NULL()

    def __evaluate_for_statement(
        self: Self, for_statement: ForStatement
    ) -> RuntimeValue:
        self.env = Environment(self.env)

        if self.env.resolve(for_statement.identifier, False) is None:
            self.env.declareVariable(for_statement.identifier, MK_NULL())

        self.env.assignVariable(
            for_statement.identifier, self.evaluate(for_statement.init)
        )

        while self.__to_algebraic_value(self.evaluate(for_statement.until)).value:
            result = self.evaluate(for_statement.body)
            if result.type == AlgebraicType.BREAK:
                break
            if result.type == AlgebraicType.RETURN:
                self.env = self.env.parent  # type: ignore
                return result
            self.evaluate(for_statement.step)

        self.env = self.env.parent  # type: ignore
        return MK_NULL()

    def __evaluate_function_statement(
        self: Self, function_statement: FunctionStatement
    ) -> RuntimeValue:
        qFunction = QFunction(function_statement, self.env)
        name = qFunction.name

        envFound = self.env.resolve(name, False)
        if envFound is None:
            self.env.declareVariable(name, MK_NULL())

        return self.env.assignVariable(name, MK_FUNCTION(qFunction))

    def __evaluate_expression(self: Self, expression: Statement) -> RuntimeValue:
        return MK_NULL()

    def __to_algebraic_value(self: Self, value: RuntimeValue) -> AlgebraicValue:
        if not hasattr(value, "value"):
            return MK_NULL()

        return value  # type: ignore
